packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -543,16 +562,25 @@ - (void)setPlaybackSpeed:(double)speed {
 }
 
 - (CVPixelBufferRef)copyPixelBuffer {
+  // Ensure video sampling at regular intervals. This function is not called at exact time intervals
+  // so CACurrentMediaTime returns irregular timestamps which causes missed video frames. The range
+  // outside of which targetTime is reset should be narrow enough to make possible lag as small as
+  // possible and at the same time wide enough to avoid too frequent resets which would lead to
+  // irregular sampling. Ideally there would be a targetTimestamp of display link used by flutter
+  // engine (FlutterTexture can provide timestamp and duration or timestamp and targetTimestamp).
+  CFTimeInterval currentTime = CACurrentMediaTime();
+  if (fabs(self.targetTime - currentTime) > self.frameUpdater.duration / 2) {
+    self.targetTime = currentTime;
+  }
+  self.targetTime += self.frameUpdater.duration;
```

Comment:
Actually as I showed earlier it is likely that rather past pixel buffers are not available (also it does not make much sense to fetch past buffers because that would mean lag so this is understandable). And because the present actually does not have any meaning (what is returned from `CACurrentMediaTime` before call to `hasNewPixelBufferForItemTime` will likely be already past during its execution), buffers need to be fetched from the future.

Another reason is that `targetTime` approximates `CADisplayLink.targetTimestamp` which is the time that represents when the next frame displays and at that time we want to show the video frame for that time (rendering of frame happens always some time behind its presentation).

Reply:
`CADisplayLink.timestamp` is the time when the last frame displayed. Selector of target object which is passed to display link is called when the system updates the screen so it should be soon after `timestamp`. `CADisplayLink.targetTimestamp` is practically `timestamp + duration`. So the goal is to track `CACurrentMediaTime + duration`.